std/algorithm/comparison.d

```diff
@@ -1565,3 +1568,196 @@ unittest
+in $(D r2). Otherwise, returns $(D false).
+
+This function can allocate if the $(D AllocateGC.yes) flag is passed. This has
+the benefit of have better complexity than the $(D AllocateGC.no) option. However,
+this will only work if the elements' equality in both ranges can be determined
+via the element's $(D toHash) method. If customized equality is needed, then the
+$(D pred) template parameter can be changed and used with the non-allocating flag.
+See $(D binaryFun) for more details on how to define $(D pred).
+
+Non allocating forward range option: $(BIGOH n^2)
+Allocating forward range option: amortized $(BIGOH r1.length) + $(BIGOH r2.length)
+
+Params:
+    pred = an optional parameter to change how equality is defined
+    r1 = A finite forward range
+    r2 = A finite forward range
+    allocate_gc = A $(D std.typecons.Flag!"AllocateGC") instance
+
+Returns:
+    $(D bool)
+*/
+
+bool isPermutation(alias pred = "a == b", Range1, Range2)
+(Range1 r1, Range2 r2, AllocateGC allocate_gc = AllocateGC.no)
+    if (isForwardRange!Range1 &&
+        isForwardRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
+    // If the two ranges aren't the same length, then they cannot
+    // possibly be permutations
+    static if (hasLength!(Range1) && hasLength!(Range2))
+    {
+        if (r1.length != r2.length)
+        {
+            return false;
+        }
+    }
+
+    alias predEquals = binaryFun!(pred);
+    alias E1 = Unqual!(ElementType!Range1);
+    alias E2 = Unqual!(ElementType!Range2);
+
+    // Skip the elements at the beginning where r1.front == r2.front,
+    // they are in the same order and don't need to be counted.
+    while (!r1.empty && !r2.empty)
+    {
+        if (predEquals(r1.front, r2.front))
+        {
+            r1.popFront;
+            r2.popFront;
+        }
+        else
+        {
+            break;
+        }
+    }
+
+    if (r1.empty && r2.empty)
+    {
+        return true;
+    }
+
+    if (allocate_gc == AllocateGC.yes)
+    {
+        int[CommonType!(E1, E2)] counts;
+
+        foreach (item; r1)
+        {
+            ++counts[item];
+        }
+
+        foreach (item; r2)
+        {
+            --counts[item];
+
+            if (counts[item] < 0)
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+    else
+    {
+        import std.algorithm.searching : count;
+
+        static if (!hasLength!(Range1) && !hasLength!(Range2))
+        {
+            // Custom walk length check to reduce the iterations as much as possible
+            long r1_length;
+
+            foreach (e; r1.save)
+            {
+                ++r1_length;
+            }
+
+            foreach (e; r2.save)
+            {
+                --r1_length;
+
+                if (r1_length < 0)
+                {
+                    return false;
+                }
+            }
```

Comment:
Almost :) Now if r1 is long and r2 isn't, then we have the same problem.

I was thinking this:
```D
auto r1s = r1.save;
auto r2s = r2.save;

while(!r1s.empty)
{
   if(r2s.empty)
      return false;
   r1s.popFront();
   r2s.popFront();
}
if(!r2s.empty)
    return false;
```